soroban-env-host/src/budget/wasmi_helper.rs

```diff
@@ -100,23 +100,47 @@ impl ResourceLimiter for Host {
     }
 }
 
-// These values are calibrated and set by us. Calibration is done with a given
-// wasmi version, and as long as the version is pinned, these values aren't
-// expected to change much.
 pub(crate) fn load_calibrated_fuel_costs() -> FuelCosts {
-    let mut fuel_costs = FuelCosts::default();
-    fuel_costs.base = 1;
-    fuel_costs.entity = 3;
-    fuel_costs.load = 2;
-    fuel_costs.store = 1;
-    fuel_costs.call = 67;
+    let fuel_costs = FuelCosts::default();
+    // Wasmi 0.36 has a simplified fuel-cost schedule, based on its new
+    // register-machine architecture. It is simply this: 1 fuel per wasm
+    // instruction, and each fuel represents moving 8 registers or 64 bytes.
+    //
+    // All this is hard-wired now (see FuelCosts::default) and it seems broadly
+    // _correct_ in terms of the actual runtime costs we see in wasmi: it costs
+    // _about_ 8-16 CPU instructions per fuel when we look at instructions we
+    // can even calibrate, and wasmi's own benchmarks suggest it runs about
+    // 8-16x slower than native code. So we will just leave their calibration
+    // as-is and hope it's not too wildly off in practice.
     fuel_costs
 }
 
-pub(crate) fn get_wasmi_config(budget: &Budget) -> Result<wasmi::Config, HostError> {
+pub(crate) fn get_wasmi_config(
+    budget: &Budget,
+    mut cmode: CompilationMode,
+) -> Result<wasmi::Config, HostError> {
     let mut config = wasmi::Config::default();
     let fuel_costs = budget.0.try_borrow_or_err()?.fuel_costs;
 
+    let enforced_limits = if cfg!(feature = "bench") {
+        // Disable limits when benchmarking, to allow large inputs.
+        EnforcedLimits::default()
+    } else {
+        let mut limits = EnforcedLimits::strict();
+        // Weaken the strict limits a bit to allow existing tests to pass.
+        // TODO: decide if it'd be better to change the tests (we need to
```

Comment:
I think it's important to make sure that no existing contracts start failing due to this. What are the tests that fail with default limits?

Reply:
You've got it a bit backwards: none of the tests fail with _default_ limits. But wasmi recently added "strict" limits that are a bit tighter, but make the VM a little less likely to be DoSable by super weird wasms.

I've decided (in this PR) to try using those strict limits; if we find people hitting them in practice we can relax them further. For the most part this is fine, but there are two limits here I'm weakening in order to make the testsuite pass:

  - There's a `min_avg_bytes_per_function` limit which .. personally I think of as possibly a bit too weird and easy to trip. It takes the number of functions and divides by the number of bytes in the wasm and if the average function size is too small, it considers the wasm probably-an-attack (the most pathological wasm is one where it's very large and made up only of empty or nearly-empty function definitions). So this "makes sense" as a limit but I think it's too strict, it prohibits lots of "trivially small wasms with just a few trivial functions" like we generate and like people playing around doing experiments are likely to generate. I just turned this limit off.
  - There's also a new limit in the strict limit that restricts the number of (data) segments to 1000. There's no particularly good reason to make a wasm with lots of segments, but we happen to have one test in the testsuite with 10000 segments to check that that's ok (because that was the previous default limit). So I left it there and turned off the segment limit. Thinking about this, I think I would prefer to just change the test. So I'll push a change that does that now.

I'll be doing tests in coming days that none of these _strict_ limits prevent replay of existing transactions on the network. I think overall it's mostly good to have them turned on.